llumnix/global_scheduler/dispatch_policy.py

```diff
@@ -16,90 +16,102 @@
+        raise NotImplementedError
+
+    def random_choice_from_top_k(self, sorted_instance_infos: List[InstanceInfo]):
+        k = min(self.topk_random_dispatch, len(sorted_instance_infos))
+        top_k_instance_infos = sorted_instance_infos[:k]
+        return random.choice(top_k_instance_infos)
 
 
 # Dispatch all requests to a single instance, used only for testing
 class Flood(DispatchPolicy):
+    def __init__(self, topk_random_dispatch: int):
+        super().__init__(topk_random_dispatch)
+
     def dispatch(self,
+                 instance_type: InstanceType,
                  instance_num_requests: Dict[str, int],
-                 available_instance_infos: List[InstanceInfo],
-                 topk_random_dispatch: int) -> str:
+                 available_instance_infos: Dict[str, InstanceInfo]) -> str:
         instance_id = max(instance_num_requests, key=instance_num_requests.get)
         return instance_id
 
 
 class Balanced(DispatchPolicy):
+    def __init__(self, topk_random_dispatch: int):
+        super().__init__(topk_random_dispatch)
+
     def dispatch(self,
+                 instance_type: InstanceType,
                  instance_num_requests: Dict[str, int],
-                 available_instance_infos: List[InstanceInfo],
-                 topk_random_dispatch: int) -> str:
-        # dispatch request according to the number of requests dispatched to instance by manager
+                 available_instance_infos: Dict[str, InstanceInfo]) -> str:
         instance_id = min(instance_num_requests, key=instance_num_requests.get)
         return instance_id
 
 
 class Load(DispatchPolicy):
+    instance_type_metric_map: Dict[InstanceType, str] = {
+        InstanceType.NO_CONSTRAINTS: 'dispatch_load_metric',
+        InstanceType.PREFILL: 'dispatch_load_metric',
+        InstanceType.DECODE: 'dispatch_load_metric',
+        InstanceType.PREFILL_AS_DECODE: 'dispatch_prefill_as_decode_load_metric',
+        InstanceType.DECODE_AS_PREFILL: 'dispatch_decode_as_prefill_load_metric'
+    }
+
+    def __init__(self, topk_random_dispatch: int):
+        super().__init__(topk_random_dispatch)
+
     def dispatch(self,
+                 instance_type: InstanceType,
                  instance_num_requests: Dict[str, int],
-                 available_instance_infos: List[InstanceInfo],
-                 topk_random_dispatch: int) -> str:
-        sorted_instance_infos = sort_instance_infos(available_instance_infos, 'dispatch_load_metric')
-        instance_info_chosen = random_choice_from_top_k(sorted_instance_infos, topk_random_dispatch)
+                 available_instance_infos: Dict[str, InstanceInfo]) -> str:
+        sorted_instance_infos = sort_instance_infos(available_instance_infos.values(),
+                                                    self.instance_type_metric_map[instance_type])
+        instance_info_chosen = self.random_choice_from_top_k(sorted_instance_infos)
```

Comment:
random_choice_from_top_k can be put to utils.py, because migration scheduler might use it.